### ISSUE TITLE
Incomplete OSM XML import

### ISSUE THREAD

**Issue by Ren-Chang** [qgis/QGIS#39358]
**Describe the bug**

Larger OpenStreetMap XML data file (for my case, it was ~100MB and contained streets and referred nodes from Overpass Turbo) cannot be loaded completely. Only part of the streets were shown, plus most of the features were not shown in attribute table (only showing 99 rows). Exporting the layer yielded only those ways with largest ~10k IDs.

**How to Reproduce**

1. Go to 'Layers > Add Layers > Add Vector Layers'
2. Open the OSM XML file and click on Add
3. Select LineString geometry type and click on OK

![20201014092344](https://user-images.githubusercontent.com/8044986/95933081-2183b280-0e00-11eb-8f00-40692a042a57.png)
[beijing-streets.zip](https://github.com/qgis/QGIS/files/5375111/beijing-streets.zip)

**QGIS and OS versions**

QGIS version
3.14.16-Pi
QGIS code revision
df27394552
Compiled against Qt
5.11.2
Running against Qt
5.11.2
Compiled against GDAL/OGR
3.0.4
Running against GDAL/OGR
3.0.4
Compiled against GEOS
3.8.1-CAPI-1.13.3
Running against GEOS
3.8.1-CAPI-1.13.3
Compiled against SQLite
3.29.0
Running against SQLite
3.29.0
PostgreSQL Client Version
11.5
SpatiaLite Version
4.3.0
QWT Version
6.1.3
QScintilla2 Version
2.10.8
Compiled against PROJ
6.3.2
Running against PROJ
Rel. 6.3.2, May 1st, 2020
OS Version
Windows 7 SP 1 (6.1)
Active python plugins
batchvectorlayersaver; 
QuickOSM; 
quick_map_services; 
db_manager; 
MetaSearch; 
processing

**Additional context**

- The OSM XML file is not corrupted because it can be correctly handled in JOSM and previous import-and-export-topology workflow in the menu of Vector > OpenStreetMap in previous QGIS versions. 
- Export of the largest way IDs mean this may not be related to 32-bit integer issue, which I also confirmed by checking the XML in a text editor to find the loaded features referred to nodes with ID larger than 4 billion.
- Running on a machine with 16GB memory. Even though the app is 32-bit, its memory footprint when importing this file was always below 300MB, probably meaning this is not a memory issue.

**Comment by gioman** [comment]
>  [beijing-streets.zip](https://github.com/qgis/QGIS/files/5375111/beijing-streets.zip)

I get nothing when loading this datasource in QGIS 3.10/3.14 on both Windows and Linux.

**Comment by Ren-Chang** [comment]
> I get nothing when loading this datasource in QGIS 3.10/3.14 on both Windows and Linux.

Yes, it seems like you got nothing, but the incomplete features are actually there if you zoom to areas around Beijing (116.4, 39.9). The empty view may be related to the empty extent and unknown feature count as shown in the layer properties, which prevented QGIS from zooming to the proper extent.

**Comment by rouault** [comment]
Using directly a large OSM XML file is impractical for performance reasons and the underlying OGR OSM driver has limitations (https://gdal.org/drivers/vector/osm.html#interleaved-reading) when using directly such files in the way a traditional driver is used by QGIS. Typically you can't get more than 100 000 features from each layer.
The most reasonable option is to use the ogr2ogr tool, which will use the driver in a special way, to convert the XML file to something that can be consumed in an efficient way, like GeoPackage.

**Comment by gioman** [comment]
> Using directly a large OSM XML file is impractical for performance reasons and the underlying OGR OSM driver has limitations (https://gdal.org/drivers/vector/osm.html#interleaved-reading) when using directly such files in the way a traditional driver is used by QGIS. Typically you can't get more than 100 000 features from each layer.
> The most reasonable option is to use the ogr2ogr tool, which will use the driver in a special way, to convert the XML file to something that can be consumed in an efficient way, like GeoPackage.

@Ren-Chang changed this to feature request.

**Comment by Ren-Chang** [comment]
> @Ren-Chang changed this to feature request.

Sounds good to me. Thank you all for the help and efforts.
